bin/stager.py

```diff
@@ -0,0 +1,343 @@
+        self.s3 = S3Agent()
+
+    def stage_file(self, target_bucket):
+        self.target = f"s3://{target_bucket}/{self.bundle.path}/{self.filename}"
+        if self._obj_is_at_target_location():
+            logger.progress(",")
+        else:
+            src_location = self.source_data_file()
+            self.copy_file_to_target_location(src_location)
+            self._delete_downloaded_file(src_location)
+        self._ensure_checksum_tags()
+
+    def _obj_is_at_target_location(self):
+        obj = self.s3.get_object(self.target)
+        if obj:
+            if obj.content_length == self.file_size:
+                logger.output("=present ")
+                return True
+            else:
+                logger.output(f"\n      exists at target but has different size: {self.file_size} / {obj.content_length}")
+        return False
+
+    def source_data_file(self):
+        location = self._find_locally() or self._download_from_source()
+        if location:
+            logger.output(f"\n      found at {location}")
+            return location
+        raise BundleMissingDataFile(f"Cannot find source for {self.filename}")
+
+    def copy_file_to_target_location(self, source_location):
+        if urlparse(source_location).scheme == 's3':
+            logger.output(f"\n      copy to {self.target} ", "C")
+            report_duration_and_rate(self.s3.copy_between_buckets,
+                                     source_location,
+                                     self.target,
+                                     size=self.file_size)
+            S3ObjectTagger(self.target).copy_tags_from_object(source_location)
+        elif urlparse(source_location).scheme == 'file':
+            local_path = urlparse(source_location).path.lstrip('/')
+            logger.output(f"\n      upload to {self.target} ", "^")
+            checksums = report_duration_and_rate(self.s3.upload_and_checksum,
+                                                 local_path,
+                                                 self.target,
+                                                 size=self.file_size)
+            S3ObjectTagger(self.target).tag_using_these_checksums(checksums)
+            logger.output("+tagging ")
+        else:
+            raise RuntimeError(f"Unrecognized scheme: {source_location}")
+
+    def _find_locally(self):
+        local_path = os.path.join(self.bundle.path, self.filename)
+        if os.path.isfile(local_path) and os.stat(local_path).st_size == self.file_size:
+            return f"file:///{local_path}"
+        return None
+
+    def _download_from_source(self):
+        logger.output(f"\n      downloading {self.original_location_url} [{sizeof_fmt(self.file_size)}]", "v")
+        dest_path = os.path.join(self.bundle.path, self.filename)
+        try:
+            report_duration_and_rate(self._download, self.original_location_url, dest_path, size=self.file_size)
+            return f"file:///{dest_path}"
```

Comment:
Yes, per https://en.wikipedia.org/wiki/File_URI_scheme the format of a file URL is `file://host/path`.   `file://localhost/` and `file:///` are equivalent.